core/blockchain.go

```diff
@@ -609,31 +610,39 @@ func (bc *BlockChain) ExportN(w io.Writer, first uint64, last uint64) error {
 	return nil
 }
 
-// insert injects a new head block into the current block chain. This method
+// writeHeadBlock injects a new head block into the current block chain. This method
 // assumes that the block is indeed a true head. It will also reset the head
 // header and the head fast sync block to this very same block if they are older
 // or if they are on a different side chain.
 //
 // Note, this function assumes that the `mu` mutex is held!
-func (bc *BlockChain) insert(block *types.Block) {
+func (bc *BlockChain) writeHeadBlock(block *types.Block) {
 	// If the block is on a side chain or an unknown one, force other heads onto it too
 	updateHeads := rawdb.ReadCanonicalHash(bc.db, block.NumberU64()) != block.Hash()
 
 	// Add the block to the canonical chain number scheme and mark as the head
-	rawdb.WriteCanonicalHash(bc.db, block.Hash(), block.NumberU64())
-	rawdb.WriteHeadBlockHash(bc.db, block.Hash())
+	// First write all header chain relative indexes
+	batch := bc.db.NewBatch()
+	rawdb.WriteCanonicalHash(batch, block.Hash(), block.NumberU64())
+	if updateHeads {
+		bc.hc.WriteHeadHeader(batch, block.Header())
```

Comment:
Also, the previous code used `block.Hash()` which remembers the hash - here it becomes recalculated. Could you make `WriteHeadHeader` take the hash as input?

Reply:
Thanks for pointing out here. You are right I can't call `WriteHeadHeader ` here directly. Although the interval is very short, but it's still better to prevent it. I will use `rawdb.WriteHeadHeaderHash(db, head.Hash())` instead

Reply:
Well, have the tests, and find if we want this guarantee that "all changes be pushed into disk first and then update the in-memory flags", then we have to break the `WriteHeadHeader` into 2 separate steps.

It's total fine, but just it's not a good approach. `WriteHeadHeader` is kind of `self-contained` function which processes all logics in the headerchain. It's weird to call  `rawdb.WriteHeadHeaderHash(db, head.Hash())` directly.